zkevm-circuits/src/pi_circuit2.rs

```diff
@@ -0,0 +1,1582 @@
+//! Use the hash value as public input.
+//!
+//! We will use three lookup tables to implement the circuit.
+//! 1. rlp_table: txs -> rlp
+//! 2. compress_table: rlp -> compress
+//! 3. hash_table: compress -> hash
+//!
+//! The dataflow:
+//! ```
+//! +----------+         +-----------+       +------------+
+//! |   txs    +---------> compress  +------->   hash     |
+//! |          |         |           |       |            |
+//! +----------+         +-----------+       +------------+
+//! ```
+
+use std::{char, marker::PhantomData};
+
+use crate::{
+    evm_circuit::util::constraint_builder::BaseConstraintBuilder,
+    witness::{Block, BlockContext},
+};
+use bytes::Bytes;
+use eth_types::geth_types::BlockConstants;
+use eth_types::sign_types::SignData;
+use eth_types::H256;
+use eth_types::{
+    geth_types::Transaction, Address, BigEndianHash, Field, ToBigEndian, ToLittleEndian, ToScalar,
+    Word,
+};
+use ethers_core::utils::keccak256;
+use halo2_proofs::{
+    circuit,
+    plonk::{Expression, Instance},
+};
+use itertools::Itertools;
+use rlp::RlpStream;
+
+use crate::table::TxFieldTag;
+use crate::table::TxTable;
+use crate::table::{BlockTable, KeccakTable};
+use crate::util::{random_linear_combine_word as rlc, Challenges, SubCircuit, SubCircuitConfig};
+use crate::witness;
+use gadgets::is_zero::IsZeroChip;
+use gadgets::util::{not, or, Expr};
+use halo2_proofs::{
+    circuit::{AssignedCell, Layouter, Region, SimpleFloorPlanner, Value},
+    plonk::{Advice, Circuit, Column, ConstraintSystem, Error, Fixed, Selector},
+    poly::Rotation,
+};
+
+/// Fixed by the spec
+const TX_LEN: usize = 10;
+const BLOCK_LEN: usize = 7 + 256;
+const EXTRA_LEN: usize = 2;
+const ZERO_BYTE_GAS_COST: u64 = 4;
+const NONZERO_BYTE_GAS_COST: u64 = 16;
+const MAX_DEGREE: usize = 10;
+
+/// Values of the block table (as in the spec)
+#[derive(Clone, Default, Debug)]
+pub struct BlockValues {
+    coinbase: Address,
+    gas_limit: u64,
+    number: u64,
+    timestamp: u64,
+    difficulty: Word,
+    base_fee: Word, // NOTE: BaseFee was added by EIP-1559 and is ignored in legacy headers.
+    chain_id: u64,
+    history_hashes: Vec<H256>,
+}
+
+/// Values of the tx table (as in the spec)
+#[derive(Default, Debug, Clone)]
+pub struct TxValues {
+    nonce: Word,
+    gas: Word, //gas limit
+    gas_price: Word,
+    from_addr: Address,
+    to_addr: Address,
+    is_create: u64,
+    value: Word,
+    call_data_len: u64,
+    call_data_gas_cost: u64,
+    tx_sign_hash: [u8; 32],
+}
+
+/// Extra values (not contained in block or tx tables)
+#[derive(Default, Debug, Clone)]
+pub struct ExtraValues {
+    // block_hash: H256,
+    state_root: H256,
+    prev_state_root: H256,
+}
```

Comment:
Ideally should be `pre_block_hash` and `block_hash`, unless we add some code to the smart contracts that allows us to put the state roots here. Also I guess the prover address needs to added to this?